cmd/sync/sync.go

```diff
@@ -0,0 +1,433 @@
+)
+
+const description = `
+Example: A sync operation dry run:
+
+  go run ./cmd/sync -n
+
+Sync runs a "merge from upstream" and submits it as a PR. This means fetching commits from an
+upstream repo and merging them into corresponding branches in a target repo. This is configured in a
+config file, by default 'eng/sync-config.json'. For each entry in the configuration:
+
+1. Fetch each SourceBranch 'branch' from 'Upstream' to a local temp repository.
+2. Fetch each 'microsoft/{branch}' from 'Target'.
+3. Merge each upstream branch 'b' into corresponding 'microsoft/b'.
+4. Push each merge commit to 'Head' (or 'Target' if 'Head' isn't specified) with a name that follows
+   the pattern 'dev/auto-merge/microsoft/{branch}'.
+5. Create a PR in 'Target' that merges the auto-merge branch. If the PR already exists, overwrite.
+   (Force push.)
+
+This script creates the temporary repository in 'eng/artifacts/' by default.
+
+To run a subset of the syncs specified in the config file, or to swap out URLs for development
+purposes, create a copy of the configuration file and point at it using a '-c' argument.
+`
+
+var dryRun = flag.Bool("n", false, "Enable dry run: do not push, do not submit PR.")
+
+var githubUser = flag.String("github-user", "", "Use this github user to submit pull requests.")
+var githubPAT = flag.String("github-pat", "", "Submit the PR with this GitHub PAT, if specified.")
+var githubPATReviewer = flag.String("github-pat-reviewer", "", "Approve the PR and turn on auto-merge with this PAT, if specified. Required, if github-pat specified.")
+
+func main() {
+	var syncConfig = flag.String("c", "eng/sync-config.json", "The sync configuration file to run.")
+	var tempGitDir = flag.String(
+		"temp-git-dir",
+		filepath.Join(getwdOrPanic(), "eng", "artifacts", "sync-upstream-temp-repo"),
+		"Location to create the temporary Git repo. A timestamped subdirectory is created to reduce chance of collision.")
+
+	var gitAuthSSH = flag.Bool("git-auth-ssh", false, "If enabled, automatically convert Target GitHub URLs into SSH format for authentication. 'git-auth-pat' is ignored if also specified.")
+	var gitAuthPAT = flag.Bool("git-auth-pat", false, "If enabled, automatically modify GitHub URLs to use 'github-user' and 'github-pat' for fetch/push access.")
+
+	buildmodel.ParseBoundFlags(description)
+
+	if *gitAuthPAT {
+		missingArgs := false
+		if *githubUser == "" {
+			fmt.Printf("Error: git-auth-pat is specified but github-user is not.")
+			missingArgs = true
+		}
+		if *githubPAT == "" {
+			fmt.Printf("Error: git-auth-pat is specified but github-pat is not.")
+			missingArgs = true
+		}
+		if missingArgs {
+			os.Exit(1)
+		}
+	}
+
+	var entries []SyncConfigEntry
+	if err := buildmodel.ReadJSONFile(*syncConfig, &entries); err != nil {
+		panic(err)
+	}
+
+	if len(entries) == 0 {
+		fmt.Printf("No entries found in config file: %v\n", *syncConfig)
+	}
+
+	currentRunGitDir, err := buildmodel.GetTempPathInDir(*tempGitDir)
```

Comment:
It's intentionally not cleaned up, so it can be inspected after a dev run gone wrong, or used to create a clone that the dev then uses to submit the automated update manually.

Maybe I shouldn't call it Temp?

Reply:
I see, it make sense to keep them then.

What about calling it `GetWorkPathInDir`?